File: BaekJoon/14888_연산자끼워넣기_BackTracking.cpp
#include <iostream>
#include <vector>
#include <limits>

using namespace std;

// 연산자끼워넣기 14888 BackTracking
// O( (N - 1)! )


int n;
vector<int> nums;   // 수열 벡터
vector<int> ops(4); // 각 연산자 개수 벡터
int minValue = numeric_limits<int>::max();  // 최소값
int maxValue = numeric_limits<int>::min();  // 최대값

// op 에 대한 연산 수행 후 반환
inline int operations(int x, int y, int op)
{
    if (op == 0) return x + y;
    if (op == 1) return x - y;
    if (op == 2) return x * y;
    return x / y;
}

void dfs(int idx, int curRes)
{
    // 모든 숫자를 썼다면
    if (idx == n)
    {
        minValue = min(minValue, curRes);
        maxValue = max(maxValue, curRes);
        return;
    }

    // 모든 연산자 탐색
    for (int i = 0; i < 4; ++i)
    {
        // 연산자가 있다면
        if (ops[i] > 0)
        {
            --ops[i];
            dfs(idx + 1, operations(curRes, nums[idx], i));
            ++ops[i]; // 백트래킹
        }
    }
}

int main(void)
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    cin >> n;
    nums.resize(n);

    for (int& num : nums) cin >> num;
    for (int& op : ops) cin >> op;

    dfs(1, nums[0]);

    cout << maxValue << "\n" << minValue;
    

    return 0;
}

File: BaekJoon/Java/14888_연산자끼워넣기_BackTracking.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

// 연산자끼워넣기 14888 BackTracking
// O( (N - 1)! )


public class Main {
        static int n;
        static int[] nums, ops;
        static int maxValue = Integer.MIN_VALUE;
        static int minValue = Integer.MAX_VALUE;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = null;

        n = Integer.parseInt(br.readLine());

        st = new StringTokenizer(br.readLine());
        nums = new int[n];
        for (int i = 0; i < n; ++i) {
            nums[i] = Integer.parseInt(st.nextToken());
        }

        st = new StringTokenizer(br.readLine());
        ops = new int[4];
        for (int i = 0; i < 4; ++i) {
            ops[i] = Integer.parseInt(st.nextToken());
        }

        dfs(1, nums[0]);

        System.out.println(maxValue + "\n" + minValue);
    }

    private static void dfs(int idx, int curRes) {
        if (idx == n) {
            minValue = Math.min(minValue, curRes);
            maxValue = Math.max(maxValue, curRes);
            return;
        }

        for (int i = 0; i < 4; ++i) {
            if (ops[i] > 0) {
                --ops[i];
                int nextRes = switch (i) {
                    case 0 -> curRes + nums[idx];
                    case 1 -> curRes - nums[idx];
                    case 2 -> curRes * nums[idx];
                    case 3 -> curRes / nums[idx];
                    default -> throw new IllegalStateException("");
                };
                dfs(idx + 1, nextRes);
                ++ops[i];
            }
        }
    }
}
